### PR TITLE
feat(cdf-ts): Add support for packages to supply typing info for components, services and relations

### PR DESCRIPTION
**Pull request by cwang**
## Description

- [x] Add support for package typing
- [x] Generate json schema for the tying file
- [x] Add tests

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor and code improvement (non-breaking change which improves code quality and/or performance)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation
- [ ] Tests
- [ ] Other chores such as maintenance

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration by modifying the list below.

- [ ] Test A
- [ ] Test B
- [ ] Test C

**Test Configuration**:

Please describe the test setup. List them below as bullet points.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings
- [ ] The commit message follows the convention of this project

### CODE REVIEW

File: docs/user-guide/framework-spec.md [hunk 1]
```diff
@@ -128,11 +142,17 @@ There can be as many Custom modules as needed.
 
 TBD
 
-## Package Spec
+## Package Definition Pack
```
Comment (janaka) [inline docs/user-guide/framework-spec.md:145]: Feels like we can avoid introducing this term. Use `Package` (for PDP) and `Package Manifest`.

Comment (cwang) [reply docs/user-guide/framework-spec.md:145]: I think PDP has its own definition, meaning a pack of files exposing package metadata (the manifest is the index file, among the other three). I've pushed an update for this part.

Comment (janaka) [reply docs/user-guide/framework-spec.md:145]: I didn't mean we should use it to describe the four files you have labelled PDP. Obviously it cannot mean two things. Can we describe the world effectively without introducing a new label?

Comment (janaka) [reply docs/user-guide/framework-spec.md:145]: It's incremental for us but for somebody new it's increasing the total complexity of learning and understanding CDF.

Comment (cwang) [reply docs/user-guide/framework-spec.md:145]: You have a point. The way I see it is that they're all integral parts of a package definition therefore grouping them together makes sense, which leads to another naming challenge :) Let me try eliminating the pack definition.

**Review by janaka — APPROVED** [review]

File: docs/user-guide/framework-spec.md (hunk 1 above)
Comment (janaka) [reply docs/user-guide/framework-spec.md:145]: idea: maybe referring to those four things _meta data_ when needed is sufficient. I was just trying to think how we might visualise in a diagram.

Comment (cwang) [reply docs/user-guide/framework-spec.md:145]: Yep it's between `definitions` and `metadata` both of which I've been using subconsciously anyway. We just need to decide on one.

File: docs/user-guide/framework-spec.md [hunk 2]
```diff
@@ -128,11 +142,74 @@ There can be as many Custom modules as needed.
...
+
+An example of the files is show below,
+
+![files in PDP](../assets/cdf-package-definition-files.png)
+
+Those files are described individually below.
+
+### Package Manifest
+
+It is aptly named `cdf.manifest.json` inside a package at root level, which contains attributes listed below to describe the package and links to other metadata files. It is the master plan when defining a package for consumption.
```
Comment (janaka) [inline docs/user-guide/framework-spec.md:157]:
```suggestion
It is aptly named `cdf.manifest.json` inside a package at root level, which contains attributes listed below to describe the package and links to other metadata files. It is the top-level plan when defining a package for consumption. 
```

File: docs/user-guide/framework-spec.md [hunk 3]
```diff
@@ -128,11 +142,74 @@ There can be as many Custom modules as needed.
...
+
+### Package Manifest
+
+It is aptly named `cdf.manifest.json` inside a package at root level, which contains attributes listed below to describe the package and links to other metadata files. It is the master plan when defining a package for consumption. 
+
+An example (as defined in `metadata.ts` in cdf-typescript) is shown below,
+
+```typescript
+/**
+ * Interface for modelling a package manifest json file that provides informations about the package.
```
Comment (janaka) [inline docs/user-guide/framework-spec.md:163]: typo:

```suggestion
 * Interface for modelling a package manifest json file that provides information about the package.
```

File: docs/user-guide/framework-spec.md [hunk 4]
```diff
@@ -156,32 +233,41 @@ Command: TEXT | REQUIRED
...
+#### Construct Level Attributes
 
 Name: SIMPLE_TEXT | REQUIRED
 Description: MARKDOWN_TEXT
 Icon: URL
 CloudVendors: TEXT_LIST | IN_LIST(pre-fined list of all cloud vendors)
 DefaultCloudVendor: TEXT | IN_LIST(pre-fined list of all cloud vendors)
 
-#### Network
+#### Types Definition via Plan JSON
```
Comment (janaka) [inline docs/user-guide/framework-spec.md:244]: I'm finding this section quite difficult to follow. I don't see the connect with the rest and how to put it together. Seems like a package author needs to understand how to create this.

**Review by janaka — CHANGES_REQUESTED** [review]
Type Defs via Plan JSON is the main point that needs addressing. See inline comment.

File: docs/user-guide/framework-spec.md [hunk 5]
```diff
@@ -190,46 +241,15 @@ Icon: URL
...
-   * Available relations between a specific component/service and another component/service.
-   * Both the key and value here refer to the entries defined in `components` and `services` fields.
-   */
-  relations: { [key: string]: string };
-}
-```
+It is a JSON schema file, usually generated with the help of `cdf-cli` tool from reading the TypeScript based package configurations. It is not expected to be manually edited afterwards. Whenever there is a change to the package configurations, the schema file should be re-generated.
 
-*The name of the types definition file can be arbitrary as long as it is referenced correctly by the `Package Manifest` file.*
+*The name of the config definition can be arbitrary as long as it is referenced correctly by the `Package Manifest` file.*
```
[the remark this answers is not in the document]
Comment (cwang) [reply docs/user-guide/framework-spec.md:252]: The `*` is the markdown formatting

File: core/typescript/src/lib/metadata.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  /**
+   * The reference to the json schema file for creating a valid json config when using the package.
+   */
+  schema: string;
+  /**
+   * Information about network.
+   */
+  network: NetworkInfo;
+  /**
+   * Available types and subtypes for components.
```
[the remark this answers is not in the document]
Comment (cwang) [reply core/typescript/src/lib/metadata.ts]: That makes sense. Due to its wide coverage, I'd leave it for another PR dedicated to renaming like this across codebase and docs.

**Comment by cwang** [comment]
> Type Defs via Plan JSON is the main point that needs addressing. See inline comment.

This is the compromise we have to make. It comes down to two choices, both of which have implications on package authors' workflow:

- Use the single json schema for not only the structure of the config file but also the available types in components/services: It is best for tools providers (us) and project creators as you can validate a config in one go. However there is no way to **generate** this single json schema; in order to achieve this, we'd ask package authors to painstakingly manually edit the generated json schema to add types into it, which is not a repeatable process either.
- Use a manifest json to keep types. We need such a manifest json file anyway to keep package metadata, and it is supposed to be manually edited and maintained by package authors. The expected workflow is that package code would read from this manifest file as single source of truth about a package, including available types to avoid duplications. It may feel like a unnatural choice (I felt the same way but I also struggled to get first choice working) but is the best compromise.

Hope it makes sense.

File: docs/user-guide/framework-spec.md (hunk 4 above)
Comment (janaka) [reply docs/user-guide/framework-spec.md:244]: if not needed lets drop. presume was an oversight that you left in? I'm not missing something right?

Comment (janaka) [reply docs/user-guide/framework-spec.md:244]: you reply to the general comment now contradicts my understanding.

File: docs/user-guide/framework-spec.md [hunk 7]
```diff
@@ -128,11 +142,74 @@ There can be as many Custom modules as needed.
...
-## Package Spec
-
-The `Package Spec` is based on the Framework spec `Types`, `Constraints`, and `Construct Types` defined above. It’s intended to serve as the interface between the high-level user-facing (App Devs) language layer and the low-level implementation layer. This leaks implementation hence not part of the framework spec at this point.
-
-### Package Level Attributes
+## Package Metadata
+
+The package metadata consist a pack of files covering all that has been mentioned above. It’s intended to be used by software tooling to understand the abstraction that a package makes available for project creators to use. The primary goal is making it a machine readable protocol so that software tooling can be developed to further aid project creators when using a selected package. 
+
+An example of the files is show below,
```
Comment (janaka) [inline docs/user-guide/framework-spec.md:149]: typo
```suggestion
An example of the files is shown below,
```

File: docs/user-guide/framework-spec.md (hunk 4 above)
Comment (janaka) [reply docs/user-guide/framework-spec.md:244]: so do we need the types in a JSON file or not?

File: docs/user-guide/framework-spec.md [hunk 8]
```diff
@@ -128,11 +142,74 @@ There can be as many Custom modules as needed.
...
+
+![Package metadata files](../assets/cdf-package-definition-files.png)
```
Comment (janaka) [inline docs/user-guide/framework-spec.md:161]: is there a reason why `metadata.ts` is not listed here?

Comment (cwang) [reply docs/user-guide/framework-spec.md:161]: This is in a package where we expect `cdf.manifest.json` - `metadata.ts` is in CDF to help package authors load this manifest file, serving as a validator if you like

**Review by janaka — APPROVED** [review]

File: docs/user-guide/framework-spec.md (hunk 4 above)
Comment (cwang) [reply docs/user-guide/framework-spec.md]: We do

**Pull request merged**